### PR TITLE
fix: security tab visibility

### PR DESCRIPTION
**Pull request by abhinavkrin**
## Proposed changes (including videos or screenshots)

This change fixes the visibility of the Security tab in user settings, ensuring that users can change their passwords even when 2FA or E2E is disabled. Previously, the option to change the password was hidden unless one of these features was enabled, which could lead to user confusion. This fix ensures that the Security tab is always visible, but it only displays relevant settings based on the user's configuration.

https://github.com/RocketChat/Rocket.Chat/assets/15830206/546321b7-eceb-4242-af91-a69dee2f2f74

## Issue(s)

This PR addresses the issue where users were unable to find the password change option due to the conditional visibility of the Security tab based on 2FA or E2E settings.
## Steps to test or reproduce

1. Disable 2FA and E2E in the workspace settings.
2. Go to My Account -> Security tab.
3. Verify that the password change option is visible and functional.

## Further comments

[SUP-458](https://rocketchat.atlassian.net/browse/SUP-458)

[SUP-458]: https://rocketchat.atlassian.net/browse/SUP-458?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by tiagoevanp — APPROVED** [review]

File: .changeset/afraid-poets-sparkle.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@rocket.chat/meteor': patch
+---
+
+Fixing Security tab visibility to allow password changes when 2FA/E2E is disabled.
```
Comment (debdutdeb) [inline .changeset/afraid-poets-sparkle.md:5]:
```suggestion
Fixed Security tab visibility to allow password changes when 2FA/E2E is disabled.
```

File: apps/meteor/client/views/account/security/AccountSecurityPage.tsx [hunk 2]
```diff
@@ -30,10 +30,11 @@ const AccountSecurityPage = (): ReactElement => {
 	const twoFactorTOTP = useSetting('Accounts_TwoFactorAuthentication_By_TOTP_Enabled');
 	const twoFactorByEmailEnabled = useSetting('Accounts_TwoFactorAuthentication_By_Email_Enabled');
 	const e2eEnabled = useSetting('E2E_Enable');
+	const allowPasswordChange = useSetting('Accounts_AllowPasswordChange');
 
 	const passwordFormId = useUniqueId();
 
-	if (!twoFactorEnabled && !e2eEnabled) {
+	if (!twoFactorEnabled && !e2eEnabled && !allowPasswordChange) {
```
Comment (dougfabris) [inline apps/meteor/client/views/account/security/AccountSecurityPage.tsx:37]:
```suggestion

```
I don't think we need this verification here, since we already do on the route

**Review by dougfabris — CHANGES_REQUESTED** [review]

**Comment by dougfabris** [comment]
@abhinavkrin only accepting my suggestion will break the code, pay attention 🚨

**Review by jessicaschelly — CHANGES_REQUESTED** [review]

File: apps/meteor/tests/e2e/access-security-page.spec.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	});
+
+	test.beforeEach(async ({ page }) => {
+		poAccountProfile = new AccountProfile(page);
+		await page.goto('/account/security');
+		await page.waitForLoadState('networkidle');
+	});
+
+	test.afterEach(async ({ api }) => {
+		if (!isFirstTestOver) {
```
[the remark this answers is not in the document]
Comment (abhinavkrin) [reply apps/meteor/tests/e2e/access-security-page.spec.ts:42]: Hey, On updating the settings, the page needs to be refreshed in order the changes to be reflected in sidenav. I moved it to afterEach to save time. As sometimes the tests used to fail due to timeout. I will try to move it to the test.

**Review by dougfabris — APPROVED** [review]
All good at frontend perspective

File: apps/meteor/tests/e2e/access-security-page.spec.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+			getSettingValueById(api, 'E2E_Enable'),
+		]);
+
+		enableSettings = async (api, { passwordChange, twofa, e2e }) => {
+			await Promise.all([
+				setSettingValueById(api, 'Accounts_AllowPasswordChange', passwordChange),
+				setSettingValueById(api, 'Accounts_TwoFactorAuthentication_Enabled', twofa),
+				setSettingValueById(api, 'E2E_Enable', e2e),
+			]);
+		}
```
Comment (MarcosSpessatto) [inline apps/meteor/tests/e2e/access-security-page.spec.ts:30]: In this case, let's use the `setSettingValueById` directly here instead of creating this function, I think it would be more explicit to see.

File: apps/meteor/tests/e2e/access-security-page.spec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (MarcosSpessatto) [reply apps/meteor/tests/e2e/access-security-page.spec.ts:42]: But from what the code is doing, we are executing this only once, after the first test case right?

In this case, I think it would be clearer to put it in a before inside the second block of the tests.

File: apps/meteor/tests/e2e/utils/test.ts [hunk 5]
```diff
@@ -13,15 +13,17 @@ const PATH_NYC_OUTPUT = path.join(process.cwd(), '.nyc_output');
...
 
+export type BaseTestAPI = {
+	recreateContext(): Promise<void>;
+	login(credentials: { username: string; password: string }): Promise<APIRequestContext>;
+	get(uri: string, params?: AnyObj, prefix?: string): Promise<APIResponse>;
+	post(uri: string, data: AnyObj, prefix?: string): Promise<APIResponse>;
+	put(uri: string, data: AnyObj, prefix?: string): Promise<APIResponse>;
+	delete(uri: string, params?: AnyObj, prefix?: string): Promise<APIResponse>;
+}
+
```
Comment (MarcosSpessatto) [inline apps/meteor/tests/e2e/utils/test.ts:24]: By following the first comment and removing the function to update the settings, we don't need this anymore

File: apps/meteor/tests/e2e/utils/test.ts [hunk 6]
```diff
@@ -13,15 +13,17 @@ const PATH_NYC_OUTPUT = path.join(process.cwd(), '.nyc_output');
...
 export type BaseTest = {
-	api: {
-		recreateContext(): Promise<void>;
-		login(credentials: { username: string; password: string }): Promise<APIRequestContext>;
-		get(uri: string, params?: AnyObj, prefix?: string): Promise<APIResponse>;
-		post(uri: string, data: AnyObj, prefix?: string): Promise<APIResponse>;
-		put(uri: string, data: AnyObj, prefix?: string): Promise<APIResponse>;
-		delete(uri: string, params?: AnyObj, prefix?: string): Promise<APIResponse>;
-	};
+	api: BaseTestAPI;
```
Comment (MarcosSpessatto) [inline apps/meteor/tests/e2e/utils/test.ts:26]: Same

File: apps/meteor/tests/e2e/utils/test.ts [hunk 7]
```diff
@@ -116,7 +118,7 @@ export const test = baseTest.extend<BaseTest>({
 	},
 	makeAxeBuilder: async ({ page }, use) => {
 		const SELECT_KNOW_ISSUES = ['aria-hidden-focus', 'nested-interactive']
-		
+
```
Comment (MarcosSpessatto) [inline apps/meteor/tests/e2e/utils/test.ts:121]:
```suggestion
```

**Review by MarcosSpessatto — CHANGES_REQUESTED** [review]

File: apps/meteor/tests/e2e/access-security-page.spec.ts [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+			setSettingValueById(api, 'Accounts_TwoFactorAuthentication_Enabled', twofaEnabledDefaultValue),
+			setSettingValueById(api, 'E2E_Enable', e2eEnabledDefaultValue),
+		]);
+	});
+
+	test('security tab is invisible when password change, 2FA and E2E are disabled', async ({ page }) => {
+		const securityTab = poAccountProfile.sidenav.linkSecurity;
+		await expect(securityTab).not.toBeVisible();
+		const mainContent = page.locator('.main-content').getByText('You are not authorized to view this page.').first();
+		await expect(mainContent).toBeVisible({ timeout: 10000 });
```
Comment (MarcosSpessatto) [inline apps/meteor/tests/e2e/access-security-page.spec.ts:61]: I think we also should rely on timeouts, @jessicaschelly can you help us here?

**Review by MarcosSpessatto — CHANGES_REQUESTED** [review]

File: apps/meteor/tests/e2e/utils/test.ts
Comment (MarcosSpessatto) [inline apps/meteor/tests/e2e/utils/test.ts:1]: Since all of these changes are only formatting stuff, let's undo the changes in this file. If we want to reformat it, let's create a separate PR for that.

**Review by MarcosSpessatto — CHANGES_REQUESTED** [review]

**Review by jessicaschelly — DISMISSED** [review]

**Review by jessicaschelly — APPROVED** [review]

**Review by MarcosSpessatto — APPROVED** [review]

**Review by tiagoevanp — APPROVED** [review]
